Propagate CLI arguments when updating

SaschaCowley opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests.

Please initially open PRs as a draft.
When you would like a review, mark the PR as "ready for review".
See https://github.com/nvaccess/nvda/blob/master/.github/CONTRIBUTING.md.
-->

### Link to issue number:

None

### Summary of the issue:

When updating NVDA, command line arguments are not passed from the running copy to the temporary copy. This may have security implications for arguments such as `-c` or `--disable-addons`.

### Description of user facing changes

The `-c`/`--config-path` and `--disable-addons` arguments are now propagated to the temporary copy of NVDA when an update is started from within NVDA.

### Description of development approach

Refactored `updateCheck._executeUpdate` to use a helper function to generate the parameters to call the launcher with. In addition to the parameters that were already passed, the following parameters are now passed:
- If `globalVars.appArgs.disableAddons` is `True`, the `--disable-addons` flag is passed.
- The `--config-path` argument is passed with the current configuration directory. We do not directly pass the path that NVDA was called with because:
  * If `-c` wasn't specified at the CLI, it will be the default config path, which is what NVDA would have defaulted to anyway.
  * If it was specified at the CLI, it will be what was specified there, as the only change we make to it is to absolutise it.
  * Sniffing `sys.argv` to find `-c` or `--config-path` adds a performance overhead (and code complexity) for no benefit, because anyone fiddling around to change `globalVars.appArgs.configPath` or `NVDAState.WritePaths.ConfigDir` can change `sys.argv` too.

### Testing strategy:

Set `buildVersion.version_major = 1`, and built a launcher with `updateVersionType=stable` (since the current stable version is 2024.2).

Created a dummy add-on that beeps at NVDA start-up.

For each of the following test cases, I installed the dummy add-on and changed the synthesiser to eSpeak so it was easy to tell if the test cases were working as expected. The test cases were performed on installed and portable copies of NVDA as created with the launcher generated above.

The following tests were performed:

- Ran `nvda.exe`. Updated NVDA and observed the add-on was run and customised settings were used.
- Ran `nvda.exe --disable-addons`. Updated NVDA and observed the add-on was not run but customised settings were used.
- Ran `nvda.exe -c c:\test` (nonexistant directory). Updated NVDA and observed the add-on was not run and default settings were used.

### Known issues with pull request:

NVDA should ideally be refactored so that there is a single source of truth for its CLI args, as they're currently repeated in multiple places (`nvda.pyw`, `globalVars.py`, `updateCheck.py` and possibly others).

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

<!-- Please keep the following -->
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Enhanced command line options for NVDA, improving update controls with respect to configuration paths and add-ons.
  - Added functionality for generating execution parameters during installation, facilitating better handling of custom configurations.

- **Bug Fixes**
  - Restored NVDA's reliance on events for caret movement, improving responsiveness and accuracy in tracking.

- **Refactor**
  - Improved error handling and parameter generation for the update process, ensuring a valid path is provided before execution.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

SaschaCowley committed:
Updated to use `executeFlags` and `executeOptions` structures, rather than strings directly, so we can set flags or options multiple times without worrying about duplicates

SaschaCowley committed:
Fixed up `_executeUpdate`

- Refactored the command line argument generation into a helper function.
- Fixed a typo where command line options were included twice and flags weren't included.

SaschaCowley committed:
Tidied

- Added type hints
- Added docstrings and comments
- Renamed some variables
- Raise `valueError` if `destPath` isn't passed to `_executeUpdate`
- Linted

SaschaCowley committed:
Merge branch 'master' into propagateCliArgs

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

SaschaCowley committed:
Added changelog entry

SaschaCowley committed:
Merge branch 'propagateCliArgs' of https://github.com/nvaccess/nvda into propagateCliArgs

SaschaCowley commented on user_docs/en/changes.md:
```diff
@@ -17,6 +17,8 @@ The available options are:
 
 ### Changes
 
+* The `-c`/`--config-path` and `--disable-addons` command line options are now respected when launching an update from within NVDA. (#16937)
+
```
@coderabbitai this comment is not relevant to this PR.

seanbudd commented on source/updateCheck.py:
```diff
@@ -237,28 +237,50 @@ def executePendingUpdate():
+	"""
 	if not destPath:
-		return
+		raise ValueError("destPath must be a non-empty string")
```
I'm concerned this will throw an unrecoverable error when updating from a version of NVDA that  doesn't send `destPath`

seanbudd commented on source/updateCheck.py:
```diff
@@ -237,28 +237,50 @@ def executePendingUpdate():
-				)
-			)
+			executeFlags.update(("--create-portable", "-m"))
+			executeOptions["--portable-path"] = f'"{portablePath}"'
```
```suggestion
			executeOptions["--portable-path"] = portablePath
```

SaschaCowley replied:
This will fail if there are spaces in `portablePath`.

seanbudd commented on source/updateCheck.py:
```diff
@@ -237,28 +237,50 @@ def executePendingUpdate():
 
 	_setStateToNone(state)
 	saveState()
+	# #4475: ensure that the new process shows its first window, by providing SW_SHOWNORMAL
```
is this comment still relevant?

SaschaCowley replied:
I don't think so. Removing.

seanbudd commented on source/updateCheck.py:
```diff
@@ -237,28 +237,50 @@ def executePendingUpdate():
+	"""
 	if not destPath:
-		return
+		raise ValueError("destPath must be a non-empty string")
```
can you explain why this has to raise rather than return?
I think it is safer to log an error here. 
I'm assuming there are existing cases calling `_executeUpdate` with `str | None` which caused this logic/handling to exist in the first place.

seanbudd replied:
If you intend to change this, can you confirm/add type checking to callers of `_executeUpdate`?

SaschaCowley replied:
I raised more out of habit than anything: this is an invalid value so `raise valueError`. I'll happily change and log an error though.

seanbudd commented on source/updateCheck.py:
```diff
@@ -237,28 +237,50 @@ def executePendingUpdate():
+		executeFlags.add("--disable-addons")
+	# pass the config path to the new instance, so that if a custom config path is in use, it will be inherited.
+	# If the default con fig path is in use, the new instance would use it anyway, so there is no harm in passing it.
+	executeOptions["--config-path"] = f'"{WritePaths.configDir}"'
```
```suggestion
	executeOptions["--config-path"] = WritePaths.configDir
```

SaschaCowley replied:
This will fail if there are spaces in `WritePaths.configDir`.

seanbudd reviewed (commented):
Approach generally looks good to me, nice work on this

michaelDCurran commented:
When the update is complete and the newly installed NVDA starts, should it also receive the same commandline parameters?
This would need to be handled in source/gui/installerGui.py, line 132 I think.

SaschaCowley commented:
@michaelDCurran
> When the update is complete and the newly installed NVDA starts, should it also receive the same commandline parameters? 

There isn't the same risk here, as the new copy will not be running as administrator. However, if you still think the arguments should be propagated, I'm happy to do so.

SaschaCowley committed:
Log error and return rather than raising. Remove unneeded comment.

seanbudd commented:
pre-commit.ci run

michaelDCurran commented:
> @michaelDCurran
> 
> > When the update is complete and the newly installed NVDA starts, should it also receive the same commandline parameters?
> 
> There isn't the same risk here, as the new copy will not be running as administrator. However, if you still think the arguments should be propagated, I'm happy to do so.
What are your thoughts on this @CyrilleB79?

LeonarddeR commented:
> There isn't the same risk here, as the new copy will not be running as administrator.

I'm puzzled. Even when updating, NVDA itself doesn't run as admin, only nvda_slave will be running elevated.

I would personally also force the command line parameters on the copy after update, that is, as long as this is also done on a normal restart using the restart dialog.

SaschaCowley commented:
> I'm puzzled. Even when updating, NVDA itself doesn't run as admin, only nvda_slave will be running elevated.
Apologies, this comment was based on a misunderstanding of mine.
> I would personally also force the command line parameters on the copy after update, that is, as long as this is also done on a normal restart using the restart dialog.
We currently propagate most arguments when restarting, with the notable exception of `--disable-addons`. I agree that, with this in mind, we should definitely propagate `-c`. The `--disable-addons` case is a pickle, however: the assumption at the moment seems to be that it is only ever wanted as a temporary state that should not be persisted. Whether or not this should be the case is out of scope for this issue, but I am slightly worried that propagating it throughout the update process may be confusing since users may expect that after NVDA restarts (which it audibly does when updating) that their add-ons will be back online. I'd appreciate your feedback on this before proceeding, @LeonarddeR , @michaelDCurran, @CyrilleB79

SaschaCowley committed:
Updated to use simpler data structures and function more similarly to `core.restart`

SaschaCowley committed:
Updated the installer to propagate -c and --disable-addons

SaschaCowley commented:
I have now updated the installer to propagate the `--disable-addons` flag and the `-c`/`--config-path` option. This turned out to be more difficult than anticipated, since the config path in use when running from the launcher may be a temporary directory. As such, I have used the value from `sys.argv` directly in this case. To test that this works as expected, I followed the build and install steps above. When checking for updates, I allowed NVDA to download the new version, then, before allowing the update to proceed, replaced the new version's launcher with the custom launcher I created (IE copy `nvda_snapshot_2024.1.exe` to `%temp`, delete `%temp%\nvda_update_XXXXXXXX.exe` and rename `%temp%\nvda_snapshot_2024.1.exe` to `%temp%\nvda_update_XXXXXXXX.exe`).

SaschaCowley commented:
@coderabbitai review

seanbudd reviewed (approved):
Nice approach for _generate_executionParameters!
